test/functional/forall/test-forall-rangesegment.hpp

```diff
@@ -0,0 +1,76 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef __TEST_FORALL_RANGESEGMENT_HPP__
+#define __TEST_FORALL_RANGESEGMENT_HPP__
+
+#include "RAJA/RAJA.hpp"
+#include "camp/resource.hpp"
+
+#include "test-forall.hpp"
+
+using namespace camp::resources;
+using namespace RAJA;
+using namespace camp;
+
+template<typename INDEX_TYPE, typename WORKING_RES, typename EXEC_POLICY>
+void ForallRangeSegmentFunctionalTest(INDEX_TYPE first, INDEX_TYPE last)
+{
+  RAJA::TypedRangeSegment<INDEX_TYPE> r1(first, last);
+  INDEX_TYPE N = r1.end() - r1.begin();
+
+  camp::resources::Resource working_res{WORKING_RES()};
+  camp::resources::Resource check_res{Host()};
+
+  INDEX_TYPE * working_array = working_res.allocate<INDEX_TYPE>(N);
+  INDEX_TYPE * check_array   = check_res.allocate<INDEX_TYPE>(N);
+  INDEX_TYPE * test_array    = check_res.allocate<INDEX_TYPE>(N);
+
+  for(INDEX_TYPE i=0; i < N; i++)
+  {
+    test_array[i]= *r1.begin() + i;
+  }
+
+  RAJA::forall<EXEC_POLICY>(r1,
+    [=] RAJA_HOST_DEVICE (INDEX_TYPE idx){
+    working_array[idx - *r1.begin()] = idx;
+  });
+
+  working_res.memcpy(check_array, working_array, sizeof(INDEX_TYPE) * N );
+
+  for(INDEX_TYPE i=0; i < N; i++)
+  {
+    ASSERT_EQ(test_array[i], check_array[i]);
+  }
+
+  working_res.deallocate(working_array);
+  check_res.deallocate(check_array);
+  check_res.deallocate(test_array);
+}
+
+
+TYPED_TEST_P(ForallFunctionalTest, RangeSegmentForall)
+{
+  using INDEX_TYPE       = typename at<TypeParam, num<0>>::type;
+  using WORKING_RESOURCE = typename at<TypeParam, num<1>>::type;
+  using EXEC_POLICY      = typename at<TypeParam, num<2>>::type;
+
+  ForallRangeSegmentFunctionalTest<INDEX_TYPE, WORKING_RESOURCE, EXEC_POLICY>(0,5);
+  ForallRangeSegmentFunctionalTest<INDEX_TYPE, WORKING_RESOURCE, EXEC_POLICY>(1,5);
+  ForallRangeSegmentFunctionalTest<INDEX_TYPE, WORKING_RESOURCE, EXEC_POLICY>(1,255);
+
+  if(std::is_signed<INDEX_TYPE>::value){
+#if !defined(__CUDA_ARCH__) && !defined(RAJA_ENABLE_TBB)
+    ForallRangeSegmentFunctionalTest<INDEX_TYPE, WORKING_RESOURCE, EXEC_POLICY>(-5,0);
```

Comment:
@mdavis36 do these tests using negative indices not work with CUDA or TBB?

Reply:
They do work but they throw warnings just like with the rangesegment unit tests.

Reply:
And we don't have a way to suppress them, correct?

Reply:
That sounds like a possible bug in our TBB and CUDA backends to me, if we allow negatives and throw warnings when they're used.  Do you happen to have the warnings handy?

Reply:
/usr/workspace/wsrzd/davis291/raja2/test/functional/forall/test-forall-rangesegment.hpp(66): warning #68-D: integer conversion resulted in a change of sign
          detected during:
            instantiation of "void gtest_suite_ForallFunctionalTest_::RangeSegmentForall<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_=camp::list<unsigned short, camp::resources::v1::Host, RAJA::policy::sequential::seq_exec>]"
/usr/WS2/davis291/raja2/blt/thirdparty_builtin/googletest-master-2020-01-07/googletest/include/gtest/internal/gtest-internal.h(470): here
            implicit generation of "gtest_suite_ForallFunctionalTest_::RangeSegmentForall<gtest_TypeParam_>::~RangeSegmentForall() [with gtest_TypeParam_=camp::list<unsigned short, camp::resources::v1::Host, RAJA::policy::sequential::seq_exec>]"
/usr/WS2/davis291/raja2/blt/thirdparty_builtin/googletest-master-2020-01-07/googletest/include/gtest/internal/gtest-internal.h(470): here
            instantiation of class "gtest_suite_ForallFunctionalTest_::RangeSegmentForall<gtest_TypeParam_> [with gtest_TypeParam_=camp::list<unsigned short, camp::resources::v1::Host, RAJA::policy::sequential::seq_exec>]"
/usr/WS2/davis291/raja2/blt/thirdparty_builtin/googletest-master-2020-01-07/googletest/include/gtest/internal/gtest-internal.h(470): here
            implicit generation of "gtest_suite_ForallFunctionalTest_::RangeSegmentForall<gtest_TypeParam_>::RangeSegmentForall() [with gtest_TypeParam_=camp::list<unsigned short, camp::resources::v1::Host, RAJA::policy::sequential::seq_exec>]"
/usr/WS2/davis291/raja2/blt/thirdparty_builtin/googletest-master-2020-01-07/googletest/include/gtest/internal/gtest-internal.h(470): here
            instantiation of class "gtest_suite_ForallFunctionalTest_::RangeSegmentForall<gtest_TypeParam_> [with gtest_TypeParam_=camp::list<unsigned short, camp::resources::v1::Host, RAJA::policy::sequential::seq_exec>]"
/usr/WS2/davis291/raja2/blt/thirdparty_builtin/googletest-master-2020-01-07/googletest/include/gtest/internal/gtest-internal.h(470): here
            [ 8 instantiation contexts not shown ]

Reply:
Yup, those are accurate warnings.  Ideally we would just make the `if` check an `if constexpr`, but we can't do that because it takes C++17.  We can probably make a workaround version in C++11 that would let you wrap it in a lambda and give the condition to a template, I'll look into that after I grab some lunch.  Alternately it could be handled with some sfinae on the ForallRangeSegmentFunctionalTest template.

Reply:
Unfortunately there's nothing to be done about a helper for C++11, for 14 I can whip something up with a bit of work, but for 11 it will likely have to be tag dispatch.  Just add a function, `runNegativeTests` or something with one variant for true_type and one for false_type and dispatch on it.